filewatcher/filewatcher.go

```diff
@@ -308,9 +360,29 @@ func New(ctx context.Context, cfg Config) (*Result, error) {
 	if err != nil {
 		return nil, err
 	}
+	isDir, err := isDirectory(cfg.Path)
+	if err != nil {
+		return nil, err
+	}
+	if isDir {
+		reader = dummyReader{
+			Path: cfg.Path,
+		}
+		// Need to walk recursively because the watcher
+		// doesnt support recursion by itself
+		err := filepath.WalkDir(cfg.Path, func(path string, info fs.DirEntry, err error) error {
+			if info.IsDir() {
+				return watcher.Add(path)
+			}
+			return nil
+		})
```

Comment:
I just remembered, there's still a corner case we haven't handled.

basically we only add the directory and all its subdirectories to watch when filewatcher itself is created, and the watch list is never updated later. here's a scenario:

1. initially filewatcher is created for `/foo/`, which contains `/foo/bar/` and `/foo/baz/` and other files
2. later, a new subdirectory, `/foo/bar3/` was added, we do not add it to the watch list, so a new file `/foo/bar3/file` being created will not trigger anything
3. `/foo/bar/` or `/foo/baz/` can also be removed later (but that likely won't cause issues in the current implementation)

I think what we need to do in `parse` callback is that if it's a directory, remove all subdirectories we added before (we likely need to record that info in the struct), and re-add all subdirectories again.

Reply:
Is this expected to happen for Vault CSI or to cover the case of directory watching in general?

Reply:
good call out
Ive added and extra block in the watcherloop to add any new sub directotories to the watcher. Ive also added a test for this case